zuko/transforms.py

```diff
@@ -629,42 +629,112 @@ class BernsteinTransform(MonotonicTransform):
     bijective = True
     sign = +1
 
-    def __init__(self, theta: Tensor, linear: bool = False, **kwargs):
+    def __init__(self, theta: Tensor, smooth_bounds: bool = False, **kwargs):
         super().__init__(None, phi=(theta,), **kwargs)
 
-        self.theta = self._increasing(theta)
-        self.linear = linear
+        self.theta = self._increasing(theta, smooth_bounds)
 
-        degree = theta.shape[-1]
-        alpha = torch.arange(1, degree + 1, device=theta.device, dtype=theta.dtype)
-        beta = torch.arange(degree, 0, -1, device=theta.device, dtype=theta.dtype)
+        self.order = self.theta.shape[-1] - 1
+        self.dtheta = self.order * (self.theta[..., 1:] - self.theta[..., :-1])
 
-        self.basis = torch.distributions.Beta(alpha, beta)
+        self.basis = self.get_bernstein_basis(self.order, device=theta.device, dtype=theta.dtype)
+        self.dbasis = self.get_bernstein_basis(
+            self.order - 1, device=theta.device, dtype=theta.dtype
+        )
+
+        # save slope on boundaries for interpolation
+        x = torch.tensor([self.eps, 1 - self.eps], device=theta.device, dtype=theta.dtype)
+        rank = self.theta.dim()
+        if rank > 1:
+            # add singleton batch dimensions
+            dims = [...] + [None] * (rank - 1)
+            x = x[dims]
+        self.offset = self.b_poly(x, self.theta, self.basis)
+        self.slope = self.b_poly(x, self.dtheta, self.dbasis)
 
     @staticmethod
-    def _increasing(theta: Tensor) -> Tensor:
+    def _increasing(unconstrained_theta: Tensor, smooth_bounds: bool) -> Tensor:
         r"""Processes the unconstrained output of the hyper-network to be increasing."""
 
-        shift = math.log(2.0) * theta.shape[-1] / 2
+        shift = math.log(2.0) * unconstrained_theta.shape[-1] / 2
+
+        theta_min = unconstrained_theta[..., :1]
+        unconstrained_theta = unconstrained_theta[..., 1:]
+
+        if smooth_bounds:
+            unconstrained_theta = torch.cat((
+                unconstrained_theta[..., :1],
+                unconstrained_theta,
+                unconstrained_theta[..., -1:],
+            ))
+        widths = torch.nn.functional.softplus(unconstrained_theta)
+
+        widths = torch.cat((theta_min, widths), dim=-1)
+
+        theta = torch.cumsum(widths, dim=-1) - shift
+
+        return theta
 
-        widths = torch.nn.functional.softplus(theta[..., 1:])
-        widths = torch.cat((theta[..., :1], widths), dim=-1)
+    @staticmethod
+    def get_bernstein_basis(order, **kwds):
```

Comment:
`_bernstein_basis`. The `_` is to prevent Sphinx from documenting the method. Also use `kwargs` instead of `kwds` and use type hints.